### PR TITLE
Add ReferenceManager to azure-core

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR adds `ReferenceManager` to `azure-core` which tracks an object reference until it is phantom reachable. Upon the referent becoming phantom reachable a cleanup action is ran, any exceptions thrown during cleanup are caught and ignored.

### CODE REVIEW

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java [hunk 1]
```diff
@@ -67,7 +68,7 @@
      * @param disableBufferCopy Determines whether deep cloning of response buffers should be disabled.
      */
     NettyAsyncHttpClient(reactor.netty.http.client.HttpClient nettyClient, boolean disableBufferCopy,
-                         long readTimeout, long writeTimeout, long responseTimeout) {
+        long readTimeout, long writeTimeout, long responseTimeout) {
```
Comment (JonathanGiles) [inline sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java:71]: fwiw, I prefer the previous style whenever I see you making these changes....the indentation makes it clearer that these are method parameters and not the first line of the method.

File: sdk/core/azure-core/src/main/java9/com/azure/core/util/ReferenceManager.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+     * @param object The object to monitor.
+     * @param cleanupAction The cleanup action to perform when the {@code object} becomes phantom reachable.
+     * @throws NullPointerException If either {@code object} or {@code cleanupAction} are null.
+     */
+    public void register(Object object, Runnable cleanupAction) {
+        cleaner.register(object, cleanupAction);
+    }
+
+    private ReferenceManager() {
+        this.cleaner = Cleaner.create();
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java9/com/azure/core/util/ReferenceManager.java:41]: I'm curious if this is the right pattern? Do we want to allow for multiple cleaners throughout our libraries (each with their own thread), or do we have a single cleaner? I would worry that by default we could end up with dozens of Cleaner instances throughout our classes.

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java:71]: > fwiw, I prefer the previous style whenever I see you making these changes....the indentation makes it clearer that these are method parameters and not the first line of the method.

Sounds good with me, we should look into having a standardized `.editorconfig` or IntelliJ code style configuration that should be a requirement when developing SDKs. Additionally, do we want to update Checkstyle rules to enforce this pattern?

File: sdk/core/azure-core/src/main/java9/com/azure/core/util/ReferenceManager.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java9/com/azure/core/util/ReferenceManager.java:41]: I went with this pattern as is mirrors what `Cleaner` itself does, but I do agree there is a good question on whether there should be a single Azure SDK reference manager thread for the entire JVM.

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java (hunk 1 above)
Comment (JonathanGiles) [reply sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClient.java:71]: I've got plans underway to standardise this.

File: sdk/core/azure-core/src/main/java9/com/azure/core/util/ReferenceManager.java (hunk 2 above)
Comment (JonathanGiles) [reply sdk/core/azure-core/src/main/java9/com/azure/core/util/ReferenceManager.java:41]: I think we need to understand the implications of doing this. I suspect we should by default consider it though.

**Comment by alzimmermsft** [comment]
Closed for #24320

**Pull request closed without merging**